File: app/src/main/java/com/example/rishabravikumar/oneapp/TimerFragment.java
package com.example.rishabravikumar.oneapp;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

public class TimerFragment extends Fragment {

    private ProgressBar progressBar;
    private TextView timerCount;
    private TextView timerTitle;
    //    private int timeInMillis = 3672000; // 1 hour 12 minutes
    private int timeInMillis = 30000; // 30 seconds
    private int hours;
    private int minutes;
    private int seconds;
    private CountDownTimer countDownTimer;

    public TimerFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_timer, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        progressBar = (ProgressBar) ((MainActivity)getActivity()).findViewById(R.id.progressBar);
        timerCount = (TextView) ((MainActivity)getActivity()).findViewById(R.id.timerCount);
        timerTitle = (TextView) ((MainActivity)getActivity()).findViewById(R.id.timerTitle);
        start();
    }

    public void start() {
        //Sets the int that represents 100% progress
        progressBar.setMax(timeInMillis/1000);

        countDownTimer = new CountDownTimer(timeInMillis, 100) {
            @Override
            public void onTick(long millisUntilFinished) {
                hours = (int)millisUntilFinished/3600000;
                minutes = ((int)millisUntilFinished - (hours*3600000))/60000;
                seconds = ((int)millisUntilFinished - (hours*3600000) - (minutes*60000))/1000;

                timerCount.setText(String.format("%02d", hours) + ":" + String.format("%02d", minutes) + ":" + String.format("%02d", seconds));

                //Sets the progress of the bar to the current amount of seconds
                progressBar.setProgress(Math.round(millisUntilFinished * 0.001f));
            }

            @Override
            public void onFinish() {
                timerTitle.setText("Done!");
                timerCount.setVisibility(View.INVISIBLE);
            }
        }.start();

        countDownTimer.start();
    }
}

File: app/src/main/java/org/hackru/oneapp/hackru/ui/main/timer/TimerFragment.kt
package org.hackru.oneapp.hackru.ui.main.timer

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Toast
import kotlinx.android.synthetic.main.fragment_timer.*
import org.hackru.oneapp.hackru.R
import org.hackru.oneapp.hackru.api.Resource
import org.hackru.oneapp.hackru.ui.main.MainActivity
import java.util.*
import kotlin.concurrent.timer

class TimerFragment : android.support.v4.app.Fragment() {
    // TAG is used with Android's Log class to organize debugging logs
    val TAG = "TimerFragment"
    private lateinit var viewModel: TimerViewModel
    private lateinit var mainActivity: MainActivity
    var countDownTimer: CountDownTimer? = null

    companion object {
        // Wondering why we use newInstance() instead of a constructor? Read this: https://stackoverflow.com/a/30867846/9968228
        fun newInstance() = TimerFragment()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mainActivity = context as MainActivity
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(R.layout.fragment_timer, container, false)

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        viewModel = ViewModelProviders.of(mainActivity).get(TimerViewModel::class.java)
        viewModel.deadline.observe(this, Observer {
            it?.let {
                when(it.state) {
                    Resource.LOADING -> {
                        // The resource is loading
                    }
                    Resource.SUCCESS -> {
                        // The resource has successfully been fetched
                        val deadline = it.data
                        val millisecondsLeft = deadline * 1000 - GregorianCalendar().timeInMillis

                        if(millisecondsLeft > 0) {
                            timer_progress.max = 86400 // 86400 seconds is 24 hours
                            countDownTimer = object : CountDownTimer(millisecondsLeft, 1000) {
                                override fun onTick(millisUntilFinished: Long) {
                                    val secondsUntilFinished = millisUntilFinished / 1000
                                    if(secondsUntilFinished < timer_progress.max) {
                                        timer_progress.progress = secondsUntilFinished.toInt()
                                        val secondsDisplay = String.format("%02d", secondsUntilFinished % 60)
                                        val minutesDisplay = String.format("%02d", (secondsUntilFinished / 60) % 60)
                                        val hoursDisplay = String.format("%02d", secondsUntilFinished / 3600)
                                        val timerCount = "$hoursDisplay:$minutesDisplay:$secondsDisplay"
                                        // We do this because it's bad practice to concatenate
                                        // strings inside TextView.setText(). If you didn't know,
                                        // timer_count.text is Kotlin shorthand for timer_count.setText()
                                        timer_count.text = timerCount
                                    } else {
                                        timer_progress.progress = timer_progress.max
                                        timer_count.text = "24:00:00"
                                    }
                                }

                                override fun onFinish() {
                                    timer_count.text = getString(R.string.timer_default_count)
                                }
                            }
                            countDownTimer?.start()
                        }
                    }
                    Resource.FAILURE -> {
                        // There was an error fetching the resource
                        Toast.makeText(context, it.msg, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        countDownTimer?.cancel()
    }

}